pybamm/models/submodels/thermal/base_thermal.py

```diff
@@ -290,6 +312,74 @@ def _current_collector_heating(self, variables):
                 Q_s_cp = self.param.p.sigma_cc * pybamm.grad_squared(phi_s_cp)
         return Q_s_cn, Q_s_cp
 
+    def _heat_of_mixing(self, variables):
+        """Compute heat of mixing source terms."""
+        param = self.param
+
+        if self.options["heat of mixing"] == "true":
+            F = pybamm.constants.F.value
+            pi = np.pi
+
+            # Compute heat of mixing in negative electrode
+            if self.options.electrode_types["negative"] == "planar":
+                Q_mix_s_n = pybamm.FullBroadcast(
+                    0, ["negative electrode"], "current collector"
+                )
+            else:
+                a_n = variables["Negative electrode surface area to volume ratio [m-1]"]
+                R_n = variables["Negative particle radius [m]"]
+                N_n = a_n / (4 * pi * R_n**2)
+                if self.x_average:
```

Comment:
Let's leave it for now, as most submodels use this anyway. When fixing the bug, it should be caught with the Ctrl+F for `if self.x_average`.

Reply:
Happy to leave it. Just a clarification about why this is different from most other submodels: In the other submodels, we are specifying whether the variable we are solving for is full or x-averaged (e.g. full or x-averaged particle concentration, if we used full concentration in SPM we'd just be solving the exact same equation 20 times instead of 1). Here, we're just evaluating an expression and returning an object of the same shape at the end, and the expression tree logic should be able to figure out how to deal with the full and x-averaged cases with the same syntax by doing things like rewrite `2 * broadcast(var)` to `broadcast(2 * var)`, but there is an edge case here that hadn't come up before

Reply:
Right, I thought it was the same case. Then happy to fix it here. Is it a matter of removing the if statement and using `c_n` and `T_n` to be the non-averaged quantities?

Reply:
Yes, it should work just always using the `x_average is False` branch

Reply:
Actually I just gave it a try and now there is an issue with dU/dsto returning 0 for the SPMe. Not sure if that's related to the other PR or not (need to dig more on it). I won't have time to take a look in the next few days, so happy to hold this until I have time or just merge it as is.

Reply:
Ah damn ok. Let's just merge as is then. My guess is that `U(broadcast(c))` becomes `broadcast(U(c))` and then it doesn't find `broadcast(c)` in the children so it thinks the derivative is zero